Function scoped fixture is run for each test, but fails all the tests with the first exception it throws itself

Hi everyone!
First of all let me thank you for such a great plugin. I use it all the time instead of the "official" one. Better loop management, focus on asyncio, timeouts out of the box - awesome!

---

Well, I ran into a problem trying to make an async fixture with funcion scope.
Let's take a look at the following script (regular pytest code, alt-pytest-asyncio is not used here).
```python
import pytest
from secrets import token_hex

OK = 0
ERROR = ~OK

def create_database(results=[OK, ERROR, OK, OK]):
    # This function emulates the DB creation.
    return results.pop(0)

@pytest.fixture(name="DB exists")
def _db() -> int:
    if create_database() == ERROR:
        print(f"Cannot create database!")
        raise Exception(token_hex(4))
    else:
        print("DB created")

pytestmark = pytest.mark.usefixtures("DB exists")

def test_0(): ...
def test_1(): ...
def test_2(): ...
def test_3(): ...
```
I create a mock DB for each test separately.
We see that the first test passes, the second fails with ERROR, and all the others pass because the fixture is called for each test as expected.
```bash
============================= test session starts ==============================
platform linux -- Python 3.11.8, pytest-8.3.5, pluggy-1.5.0 -- /root/altissue/.venv/bin/python3.11
cachedir: .pytest_cache
rootdir: /root/altissue
plugins: alt-pytest-asyncio-0.9.2
collected 4 items

repro.py::test_0 DB created
PASSED
repro.py::test_1 Cannot create database!
ERROR
repro.py::test_2 DB created
PASSED
repro.py::test_3 DB created
PASSED

==================================== ERRORS ====================================
___________________________ ERROR at setup of test_1 ___________________________
repro.py:15: in _db
    raise Exception(token_hex(4))
E   Exception: 2e14e1e4
=========================== short test summary info ============================
ERROR repro.py::test_1 - Exception: 2e14e1e4
========================== 3 passed, 1 error in 0.02s ==========================
```
I see the same thing when I have async tests with a regular sync fixture.
But if the fixture is asynchronous, like this:
```python
import pytest
from secrets import token_hex

OK = 0
ERROR = ~OK

def create_database(results=[OK, ERROR, OK, OK]):
    # This function emulates the DB creation.
    return results.pop(0)

@pytest.fixture(name="DB exists")
async def _db() -> int:
    if create_database() == ERROR:
        print(f"Cannot create database!")
        raise Exception(token_hex(4))
    else:
        print("DB created")

pytestmark = pytest.mark.usefixtures("DB exists")

async def test_0(): ...
async def test_1(): ...
async def test_2(): ...
async def test_3(): ...
```
I see the following output
```bash
============================= test session starts ==============================
platform linux -- Python 3.11.8, pytest-8.3.5, pluggy-1.5.0 -- /root/altissue/.venv/bin/python3.11
cachedir: .pytest_cache
rootdir: /root/altissue
plugins: alt-pytest-asyncio-0.9.2
collected 4 items

repro.py::test_0 DB created
PASSED
repro.py::test_1 Cannot create database!
ERROR
repro.py::test_2 DB created
ERROR
repro.py::test_3 DB created
ERROR

==================================== ERRORS ====================================
___________________________ ERROR at setup of test_1 ___________________________
repro.py:15: in _db
    raise Exception(token_hex(4))
E   Exception: 278998f4
___________________________ ERROR at setup of test_2 ___________________________
repro.py:15: in _db
    raise Exception(token_hex(4))
E   Exception: 278998f4
___________________________ ERROR at setup of test_3 ___________________________
repro.py:15: in _db
    raise Exception(token_hex(4))
E   Exception: 278998f4
=========================== short test summary info ============================
ERROR repro.py::test_1 - Exception: 278998f4
ERROR repro.py::test_2 - Exception: 278998f4
ERROR repro.py::test_3 - Exception: 278998f4
========================= 1 passed, 3 errors in 0.03s ==========================
```
In other words, the fixture is called for every test (remember, this is a function scope), but if it fails with an exception, it caches the exception and propagates it to all other tests.  See the hash I added. It's all the same exception from the first failed test.
So the fixture is called for the third test, and everything is fine with the fixture, but the test still fails with the old error.

---

I digged down into the code and found out that the root cause is that when you wrap an async fixture in a sync one, you only create the async_timeout object once. The very first exception of the fixture is set as async_timeout.error and then used over and over again.
Is this a known issue? Or am I missing something when using the plugin?
This looks like a bug to me because the pytest code behaves differently.

--- a/alt_pytest_asyncio/base.py
+++ b/alt_pytest_asyncio/base.py
@@ -1,6 +1,6 @@
 import abc
 from collections.abc import Callable
-from typing import NoReturn
+from typing import NoReturn, Protocol
 
 
 class AsyncTimeout(abc.ABC):
@@ -15,6 +15,10 @@ class AsyncTimeout(abc.ABC):
 
     @abc.abstractmethod
     def raise_maybe(self, func: Callable[..., object]) -> None: ...
+
+
+class AsyncTimeoutMaker(Protocol):
+    def __call__(self) -> AsyncTimeout: ...
 
 
 class AsyncTimeoutProvider(abc.ABC):

--- a/alt_pytest_asyncio/converter.py
+++ b/alt_pytest_asyncio/converter.py
@@ -66,7 +66,9 @@ class Converter:
             return
 
         if hasattr(fixturedef.func, "__alt_asyncio_pytest_converted__"):
-            return
+            fixturedef.func = fixturedef.func.__alt_asyncio_pytest_original__  # type: ignore[misc,attr-defined]
+
+        original = fixturedef.func
 
         if inspect.iscoroutinefunction(fixturedef.func):
             async_timeout_maker = self._get_async_timeout_maker(
@@ -87,6 +89,7 @@ class Converter:
             self._convert_sync_fixture(fixturedef)
 
         fixturedef.func.__alt_asyncio_pytest_converted__ = True  # type: ignore[attr-defined]
+        fixturedef.func.__alt_asyncio_pytest_original__ = original  # type: ignore[attr-defined]
 
     def convert_pyfunc(self, pyfuncitem: pytest.Function) -> None:
         if inspect.iscoroutinefunction(pyfuncitem.obj):

--- a/alt_pytest_asyncio/converter.py
+++ b/alt_pytest_asyncio/converter.py
@@ -69,12 +69,16 @@ class Converter:
             return
 
         if inspect.iscoroutinefunction(fixturedef.func):
-            async_timeout = self._get_async_timeout(request.scope, request.getfixturevalue)
-            self._convert_async_coroutine_fixture(fixturedef, request, async_timeout)
+            async_timeout_maker = self._get_async_timeout_maker(
+                request.scope, request.getfixturevalue
+            )
+            self._convert_async_coroutine_fixture(fixturedef, request, async_timeout_maker)
 
         elif inspect.isasyncgenfunction(fixturedef.func):
-            async_timeout = self._get_async_timeout(request.scope, request.getfixturevalue)
-            self._convert_async_gen_fixture(fixturedef, request, async_timeout)
+            async_timeout_maker = self._get_async_timeout_maker(
+                request.scope, request.getfixturevalue
+            )
+            self._convert_async_gen_fixture(fixturedef, request, async_timeout_maker)
 
         elif inspect.isgeneratorfunction(fixturedef.func):
             self._convert_sync_gen_fixture(fixturedef)
@@ -89,12 +93,13 @@ class Converter:
             _obj: Any = pyfuncitem.obj
             func: Callable[..., Awaitable[object]] = _obj
 
-            async_timeout = self._get_async_timeout(
+            async_timeout_maker = self._get_async_timeout_maker(
                 "function", pyfuncitem._request.getfixturevalue
             )
 
             @wraps(func)
             def run_test(*args: object, **kwargs: object) -> object:
+                async_timeout = async_timeout_maker()
                 res = self._run(async_timeout, func, args, kwargs)
                 async_timeout.raise_maybe(func)
                 return res
@@ -108,7 +113,7 @@ class Converter:
         self,
         fixturedef: pytest.FixtureDef[object],
         request: pytest.FixtureRequest,
-        async_timeout: base.AsyncTimeout,
+        async_timeout_maker: base.AsyncTimeoutMaker,
     ) -> None:
         """
         Run our async fixture in our event loop and capture the error from
@@ -121,6 +126,7 @@ class Converter:
         def run_fixture(*args: object, **kwargs: object) -> object:
             __tracebackhide__ = True
 
+            async_timeout = async_timeout_maker()
             res = self._run(async_timeout, func, args, kwargs)
             async_timeout.raise_maybe(func)
             return res
@@ -131,7 +137,7 @@ class Converter:
         self,
         fixturedef: pytest.FixtureDef[object],
         request: pytest.FixtureRequest,
-        async_timeout: base.AsyncTimeout,
+        async_timeout_maker: base.AsyncTimeoutMaker,
     ) -> None:
         """
         Return the yield'd value from the generator and ensure the generator is
@@ -143,6 +149,8 @@ class Converter:
         @wraps(generator)
         def run_fixture(*args: object, **kwargs: object) -> object:
             __tracebackhide__ = True
+
+            async_timeout = async_timeout_maker()
 
             if "async_timeout" in kwargs:
                 kwargs["async_timeout"] = async_timeout
@@ -258,9 +266,9 @@ class Converter:
 
         return loop.run_until_complete(task)
 
-    def _get_async_timeout(
+    def _get_async_timeout_maker(
         self, scope: str, getfixturevalue: Callable[[str], object]
-    ) -> base.AsyncTimeout:
+    ) -> base.AsyncTimeoutMaker:
         assert scope in _PytestScopes
 
         default_timeout: float = 5
@@ -281,4 +289,4 @@ class Converter:
         async_timeout_provider = getfixturevalue("async_timeout")
         assert isinstance(async_timeout_provider, base.AsyncTimeoutProvider)
 
-        return async_timeout_provider.load(default_timeout=default_timeout)
+        return lambda: async_timeout_provider.load(default_timeout=default_timeout)

--- a/alt_pytest_asyncio/converter.py
+++ b/alt_pytest_asyncio/converter.py
@@ -171,11 +171,10 @@ class Converter:
                     __tracebackhide__ = True
 
                     async_timeout.use_default_timeout()
-                    await self._async_runner(async_timeout, gen_obj.__anext__, (), {})
-                    if async_timeout.error is not None:
-                        if isinstance(async_timeout.error, StopAsyncIteration):
-                            async_timeout.error = None
-                    else:
+                    if not isinstance(async_timeout.error, StopAsyncIteration):
+                        await self._async_runner(async_timeout, gen_obj.__anext__, (), {})
+
+                    if async_timeout.error is None:
                         async_timeout.error = ValueError(
                             "Async generator fixture should only yield once"
                         )

--- a/alt_pytest_asyncio/plugin.py
+++ b/alt_pytest_asyncio/plugin.py
@@ -158,7 +158,7 @@ class LoadedAsyncTimeout(base.AsyncTimeout):
             if self.error:
                 raise self.error
 
-        if self.error:
+        if self.error and not isinstance(self.error, StopAsyncIteration):
             # Use a separate function so when --tb=short is not set we don't get
             # this entire function in the output
             raise_error()

--- a/docs/index.rst
+++ b/docs/index.rst
@@ -25,6 +25,12 @@ generator fixtures.
 Changelog
 ---------
 
+.. _release-0.9.3:
+
+0.9.3 - TBD
+    * Fixed a bug where fixture errors would persist across multiple tests
+      https://github.com/delfick/alt-pytest-asyncio/issues/26
+
 .. _release-0.9.2:
 
 0.9.2 - 23 February 2025

--- a/helpers/alt_pytest_asyncio_test_driver/examples/example_fixture_failures/expected
+++ b/helpers/alt_pytest_asyncio_test_driver/examples/example_fixture_failures/expected
@@ -1,5 +1,5 @@
 =======* ERRORS =====*
-_______* ERROR at setup of test_fails_on_fixture_returns _____*
+_* ERROR at setup of test_fails_on_fixture_returns _*
 *test_fails.py:17: in fixture_returns
     await one*()
 *test_fails.py:8: in one
@@ -7,7 +7,7 @@ _______* ERROR at setup of test_fails_on_fixture_returns _____*
 *test_fails.py:12: in two
     raise ValueError*("WAT")
 E   ValueError: WAT
-_______* ERROR at setup of test_fails_on_fixture_yields _____*
+_* ERROR at setup of test_fails_on_fixture_yields _*
 *test_fails.py:22: in fixture_yields
     yield await one*()
 *test_fails.py:8: in one
@@ -15,7 +15,7 @@ _______* ERROR at setup of test_fails_on_fixture_yields _____*
 *test_fails.py:12: in two
     raise ValueError*("WAT")
 E   ValueError: WAT
-_______* ERROR at teardown of test_fails_on_fixture_fails_in_finally _____*
+_* ERROR at teardown of test_fails_on_fixture_fails_in_finally _*
 *test_fails.py:30: in fixture_fails_in_finally
     await one*()
 *test_fails.py:8: in one
@@ -23,4 +23,28 @@ _______* ERROR at teardown of test_fails_on_fixture_fails_in_finally _____*
 *test_fails.py:12: in two
     raise ValueError*("WAT")
 E   ValueError: WAT
-=======* 1 passed, 3 error* in * ===*
+_* ERROR at setup of test_fails_fixture_failure_does_not_linger_first_iterator _*
+example_fixture_failures/test_fails.py:45: in fixture_fails_once_iterator
+    yield await one()
+example_fixture_failures/test_fails.py:8: in one
+    await two()
+example_fixture_failures/test_fails.py:12: in two
+    raise ValueError*("WAT")
+E   ValueError: WAT
+_* ERROR at setup of test_fails_fixture_failure_does_not_linger_first_result _*
+example_fixture_failures/test_fails.py:55: in fixture_fails_once_result
+    await one()
+example_fixture_failures/test_fails.py:8: in one
+    await two()
+example_fixture_failures/test_fails.py:12: in two
+    raise ValueError*("WAT")
+E   ValueError: WAT
+_* ERROR at teardown of test_fails_fixture_failure_does_not_linger_first_finally _*
+example_fixture_failures/test_fails.py:68: in fixture_fails_once_finally
+    await one()
+example_fixture_failures/test_fails.py:8: in one
+    await two()
+example_fixture_failures/test_fails.py:12: in two
+    raise ValueError*("WAT")
+E   ValueError: WAT
+=======* 5 passed, 6 error* in * ===*

--- a/helpers/alt_pytest_asyncio_test_driver/examples/example_fixture_failures/test_fails.py
+++ b/helpers/alt_pytest_asyncio_test_driver/examples/example_fixture_failures/test_fails.py
@@ -30,6 +30,44 @@ async def fixture_fails_in_finally() -> AsyncGenerator[int]:
         await one()
 
 
+_done_iterator: bool = False
+_done_result: bool = False
+_done_finally: bool = False
+
+
+@pytest.fixture()
+async def fixture_fails_once_iterator() -> AsyncGenerator[None]:
+    global _done_iterator
+    if _done_iterator:
+        return
+
+    _done_iterator = True
+    yield await one()
+
+
+@pytest.fixture()
+async def fixture_fails_once_result() -> None:
+    global _done_result
+    if _done_result:
+        return
+
+    _done_result = True
+    await one()
+
+
+@pytest.fixture()
+async def fixture_fails_once_finally() -> AsyncGenerator[None]:
+    try:
+        yield
+    finally:
+        global _done_finally
+        if _done_finally:
+            return
+
+        _done_finally = True
+        await one()
+
+
 def test_fails_on_fixture_returns(fixture_returns: int) -> None:
     pass
 
@@ -39,4 +77,40 @@ def test_fails_on_fixture_yields(fixture_yields: int) -> None:
 
 
 def test_fails_on_fixture_fails_in_finally(fixture_fails_in_finally: int) -> None:
+    pass
+
+
+def test_fails_fixture_failure_does_not_linger_first_iterator(
+    fixture_fails_once_iterator: None,
+) -> None:
+    pass
+
+
+def test_fails_fixture_failure_does_not_linger_second_iterator(
+    fixture_fails_once_iterator: None,
+) -> None:
+    pass
+
+
+def test_fails_fixture_failure_does_not_linger_first_result(
+    fixture_fails_once_result: None,
+) -> None:
+    pass
+
+
+def test_fails_fixture_failure_does_not_linger_second_result(
+    fixture_fails_once_result: None,
+) -> None:
+    pass
+
+
+def test_fails_fixture_failure_does_not_linger_first_finally(
+    fixture_fails_once_finally: None,
+) -> None:
+    pass
+
+
+def test_fails_fixture_failure_does_not_linger_second_finally(
+    fixture_fails_once_finally: None,
+) -> None:
     pass

